### PR TITLE
Disable Hibernate entity scanning for default JPA setup

### PR DESCRIPTION
**Pull request by ayudovin**
this [enhancement](https://github.com/spring-projects/spring-boot/issues/15321)

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java [hunk 1]
```diff
@@ -56,6 +57,8 @@
 	 */
 	private Boolean useNewIdGeneratorMappings;
 
+	private String archiveScanner;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: That's not really what I mind for this. I think that it is safer to disable that behaviour if no property is set so we shouldn't expose a property for this at all.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java [hunk 2]
```diff
@@ -121,6 +133,12 @@ else if (!result.containsKey(AvailableSettings.USE_NEW_ID_GENERATOR_MAPPINGS)) {
 		}
 	}
 
+	private void applyArchiveScanner(Map<String, Object> result) {
+		if (!StringUtils.isEmpty(this.archiveScanner)) {
+			result.put(AvailableSettings.SCANNER, this.archiveScanner);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: If the property is not set by the user, we should set it with the disabled scanner implementation.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/orm/jpa/HibernatePropertiesTests.java [hunk 3]
```diff
@@ -123,6 +123,15 @@ public void useNewIdGeneratorMappingsFalse() {
 										"false")));
 	}
 
+	@Test
+	public void useArchiveScanner() {
+		this.contextRunner.withPropertyValues(
+				"spring.jpa.hibernate.archive-scanner:org.hibernate.boot.archive.scan.internal.DisabledScanner")
+				.run(assertHibernateProperties((hibernateProperties) -> assertThat(
+						hibernateProperties).containsEntry(AvailableSettings.SCANNER,
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/orm/jpa/HibernatePropertiesTests.java]: We need two tests here. One where we show this is set by default and one where we customize the property (for instance setting the standard implementation explicitly and then asserting we're using that.

Comment (ayudovin) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/orm/jpa/HibernatePropertiesTests.java]: Added test for default implementation

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java (hunk 2 above)
Comment (ayudovin) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: set default value for property as disable scanning

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java [hunk 4]
```diff
@@ -56,6 +57,13 @@
 	 */
 	private Boolean useNewIdGeneratorMappings;
 
+	/**
+	 * Archive Scanner. This is actually a shortcut for the "hibernate.archive.scanner"
+	 * property. When not specified will default to
+	 * "org.hibernate.boot.archive.scan.internal.DisabledScanner".
+	 */
+	private String archiveScanner = "org.hibernate.boot.archive.scan.internal.DisabledScanner";
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: Thanks for the update but that's not what I meant. There shouldn't be any property at all.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java [hunk 5]
```diff
@@ -121,6 +138,10 @@ else if (!result.containsKey(AvailableSettings.USE_NEW_ID_GENERATOR_MAPPINGS)) {
 		}
 	}
 
+	private void applyArchiveScanner(Map<String, Object> result) {
+		result.put(AvailableSettings.SCANNER, this.archiveScanner);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: That's not checking if a user has set the property

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java (hunk 4 above)
Comment (ayudovin) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: property was removed

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java (hunk 5 above)
Comment (ayudovin) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateProperties.java]: added checking

**Comment by snicoll** [comment]
@ayudovin that change broke Hibernate 5.2 support as indicated by the build failure. Have you noticed it?

**Comment by ayudovin** [comment]
@snicoll, yes, I try to find solution. Hibernate 2 doesn't have DisabledScanner and I haven't found a solution yet

**Comment by snicoll** [comment]
Thanks @ayudovin. I've [polished your contribution](https://github.com/spring-projects/spring-boot/commit/d0811b48df6ff2d860b96f6540d45f9de53aecd6) and fixed the Hibernate 5.2 problem as well.

**Comment by odrotbohm** [comment]
I wonder if it makes sense to provide a simpler boolean flag as user facing property to avoid Hibernate internals leaking into user application configuration?

**Comment by snicoll** [comment]
@odrotbohm there is not configuration exposed at all so I am not sure what you mean there. If you don't specify anything, we use the disabled scaner impl (which, I believe, matches your original report).

**Comment by odrotbohm** [comment]
Oh, and does it make sense to use the disabled scanner by default. I guess an average Boot user doesn’t expect any Hibernate-driven scanning to run in the first place anyway?

**Comment by snicoll** [comment]
> Oh, and does it make sense to use the disabled scanner by default. 

Is that a question? If that is, you lost me as you created an issue to request that scanning to be disabled by default. Unless I misunderstood it. If disabling the scanner is not 100% safe, we should revert this and clarify the scope of the issue you've created.

**Comment by odrotbohm** [comment]
Okay, maybe the changes view is messed up then. I see tests that set a pretty complicated property that directly configured the scanner. Given our experience with Hibernate those things change quite frequently so that I wondered whether something like `.…disable-hibernate-scanning` to either be `true` or `false` allow us to mitigate potential changes.

**Comment by odrotbohm** [comment]
I was just looking at what the changes tab provided on mobile view and that doesn’t seem to contain anything hinting at what default you chose.

**Comment by snicoll** [comment]
There is nothing in your request that indicated this should be configurable. We can introduce a property but that doesn't match the description of #15321. I've already indicated that if disabling the scanner in the context of a Spring Boot auto-configured app wasn't safe, I think the scope of the issue should be clarified (with a potential reopen).

As for the "pretty complicated" property, it's just the standard Hibernate property with a prefix to help us grab it.

**Pull request closed without merging**